src/vision/status_quo/carrie_tries_a_socket_sink.md

```diff
@@ -0,0 +1,113 @@
+
+```rust
+use futures::{SinkExt, StreamExt};
+use async_std::sync::{Arc, Mutex};
+use log::{debug, info, warn};
+
+/// In the connection handler:
+let (ws_sender, mut ws_receiver) = ws_stream.split();
+let ws_sender = Arc::new(Mutex::new(ws_sender));
+
+while let Some(msg) = ws_receiver.next().await {
+    debug!("Received new WS RPC message: {:?}", msg);
+
+    // Echo the request:
+    match ws_sender.lock().await.send_string(msg).await {
+        Ok(_) => {
+            info!("New WS data sent.");
+        }
+        Err(_) => {
+            warn!("WS connection closed.");
+        }
+    };
+}
+```
+
+This is necessary because both sides of the stream are stateful Iterables, if they're kept as one, the lock on the receiver couldn't be released within the loop. He's seen this pattern used in other projects in the Rust community, but is frustrated to find that the Sink trait wasn't implemented in the WebSockets middleware library he's using.
+
+Alan also tries creating a sort of poller worker thread using an intermediary messaging channel, but he has trouble reasoning about the code and wasn't able to get it to compile:
+
+```rust
+use async_std::channel;
+use async_std::sync::{Arc, Mutex};
+use log::{debug, info, warn};
+
+/// In the connection handler:
+let (ws_sender, mut ws_receiver) = channel::unbounded::<String>();
+let ws_receiver = Arc::new(ws_receiver);
+
+let ws_stream = Arc::new(Mutex::new(ws_stream));
+let poller_ws_stream = ws_stream.clone();
+
+async_std::task::spawn(async move {
+    while let Some(msg) = ws_receiver.next().await {
+        match poller_ws_stream.lock().await.send_string(msg).await {
+            Ok(msg) => {
+                info!("New WS data sent. {:?}", msg);
+            }
+            Err(msg) => {
+                warn!("WS connection closed. {:?}", msg);
+            }
+        };
+    }
+});
+
+while let Some(msg) = ws_stream.lock().await.next().await {
+    // Echo the request?
+    ws_sender.send(msg);
+}
+```
+
+Alan wonders if he's thinking about it wrong, but the solution isn't as obvious as his earlier Sink approach. Looking around, he realizes a solution to his problems already exists-- as others have been in his shoes before-- within two other nearly-identical pull requests, but they were both closed by the project maintainers. He tries opening a third one with the same code, pointing to an example where it was actually found to be useful. To his joy, his original approach works with the code in the closed pull requests in his local copy! Alan's branch is able to compile for the first time.
+
+However, almost immediately, his request was also shut down, and the maintainer suggested he try the complex, unobvious, and unspecific solution that he had already tried and just couldn't get it to work.
```

Comment:
```suggestion
However, his pull request gets shut down instantaneously, and the maintainer suggests he try the complex, unobvious, and unspecific solution that he had already tried and couldn't get it to work.
```

Reply:
I think instantaneously is the wrong phrase here, but the other suggestions are great!